src/content_scripts/applyColorSchemeJs.js

```diff
@@ -22,109 +51,294 @@ const ADDON_FAKED_WARNING = "matchMedia has been faked/overwritten by add-on web
-    } else if (MEDIA_QUERY_PREFER_COLOR[COLOR_STATUS.NO_PREFERENCE]) {
+    } else if (MEDIA_QUERY_PREFER_COLOR[COLOR_STATUS.NO_PREFERENCE].test(mediaQueryString)) {
         return COLOR_STATUS.NO_PREFERENCE;
     } else {
         return null;
     }
 }
 
 /**
- * Returns a fake MediaQueryList as best as possible.
+ * Whether a string contains media query for color scheme
  *
  * @private
- * @param {string} mediaQueryString the original media query string
- * @param {COLOR_STATUS} askedFor
- * @returns {Object} looks like MediaQueryList
+ * @param {string} mediaQueryString
+ * @returns {boolean}
  */
-function fakeMediaQueryResult(mediaQueryString, askedFor) {
-    let matches = false;
+function hasColorMediaQuery(mediaQueryString) {
+    return (mediaQueryString.includes(MEDIA_QUERY_COLOR_SCHEME));
+}
 
-    // only return true, if the asked status is the same as the one we want to fake
-    if (fakedColorStatus === askedFor) {
-        matches = true;
+/**
+ * Evaluate a media query string.
+ * Returns null if the query has nothing to do with color.
+ *
+ * @private
+ * @param {string} mediaQueryString
+ * @returns {boolean|null}
+ */
+function evaluateMediaQuery(mediaQueryString) {
+    if (fakedColorStatus === COLOR_STATUS.NO_OVERWRITE) {
+        return matchMedia(mediaQueryString).matches;
     }
 
-    // get some parts of real object
-    const realQuery = matchMedia("(prefers-color-scheme: no-preference)");
+    let requestedMedia = getColorTypeFromMediaQuery(mediaQueryString);
+    if (requestedMedia === null) {
+        return null;
+    }
 
-    return {
-        media: mediaQueryString,
-        matches: matches,
-        onchange: realQuery.onchange || null, // real value was null anyway in my tests
-        addListener: (...args) => {
-            console.warn(ADDON_FAKED_WARNING);
+    return (fakedColorStatus === requestedMedia);
+}
 
-            return realQuery.addListener(...args);
+/**
+ * Keep track of listener addition
+ * Creates hook function if necessary
+ *
+ * @private
+ * @param {function} func original listener
+ * @param {MediaQueryList} mediaQueryList
+ * @param {boolean} isOnChange
+ * @returns {function} hook
+ */
+function trackOnListener(func, mediaQueryList, isOnChange) {
+    let entry = wmFuncToEntry.get(func);
 
-            // TODO: if we really overwrite it and not exitend it, we get the error:
-            // TypeError: 'addListener' called on an object that does not implement interface MediaQueryList.
-            // …because we only fake/create an object, not a real MediaQueryList
-        },
-        removeListener: (...args) => {
-            console.warn(ADDON_FAKED_WARNING);
+    let hook, setMediaQueryList, setMediaQueryListOnChange;
+    if (!entry) {
+        hook = makeListenerHook(func);
+        setMediaQueryList = new Set();
+        setMediaQueryListOnChange = new Set();
+        wmFuncToEntry.set(func, { hook, setMediaQueryList, setMediaQueryListOnChange });
```

Comment:
Hmm this Object deconstruction is AFAIK a newer JS feature than what the oldest Firefox version we support, supports.

See here for the minimal Firefox version that should be supported:
https://github.com/rugk/website-dark-mode-switcher/blob/7c311125b670ed603983f5b56cac26dae36edb6d/src/manifest.json#L76 

Note that we could discuss bumping this of course, if it is too hard to implement to a newer ESR version, at least. It should, at minimum, however, support the latest still stable ESR Firefox version.

Reply:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer

I think object destructuring assignment and shorthand initializer are all implemented well before 67.0a1
But changed it to the older syntax.